staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -264,19 +450,216 @@ func (q *delayingType) waitingLoop() {
+func (q *delayingType) calculateNextReadyAt(w *waitForPriorityQueue) {
+	if w.Len() == 0 {
+		// remove timer when the queue becomes empty
+		if q.nextReadyHead != nil {
+			q.nextReadyAtTimerCh = make(<-chan time.Time)
+			if q.nextReadyAtTimer != nil {
+				q.nextReadyAtTimer.Stop()
+				q.nextReadyAtTimer = nil
+			}
+			q.nextReadyHead = nil
+		}
+		return
+	}
+
+	// Keep the existing nextReadyAt timer if the head duration has not changed.
+	entry := w.Peek().(*waitFor)
+	if q.nextReadyHead != nil && q.nextReadyHead.readyAt == entry.readyAt {
+		return
+	}
+
+	// Recalculate the nextReadyAtTimer
+	q.nextReadyAtTimerCh = make(<-chan time.Time)
+	if q.nextReadyAtTimer != nil {
+		q.nextReadyAtTimer.Stop()
+		q.nextReadyAtTimer = nil
+	}
+
+	now := q.clock.Now()
+	q.nextReadyAtTimer = q.clock.NewTimer(entry.readyAt.Sub(now))
+	q.nextReadyAtTimerCh = q.nextReadyAtTimer.C()
+	q.nextReadyHead = entry
+	if q.logger != nil {
+		q.log(fmt.Sprintf("waitingLoop: calculated NEXT-READY-AT timer " + entry.readyAt.Sub(now).String()))
+	}
+}
+
+// closeAllWaitForFills loops through all the channels in waitForQueueFilled and closes them before resetting the list.
+func (q *delayingType) closeAllWaitForFills() {
+	if len(q.waitForQueueFilled) > 0 {
+		if q.logger != nil {
+			q.log(fmt.Sprintf("waitingLoop: closing %d waitForQueueFilled channels", len(q.waitForQueueFilled)))
+		}
+		for _, ch := range q.waitForQueueFilled {
+			close(ch)
+		}
+		q.waitForQueueFilled = make([]chan int, 0)
+	}
+}
+
+// processArrivingMessage decides whether to accept or drop an entry based on its waitFor settings and state of the queue.
+// It implements the following behaviours:
+//   - An entry is dropped if it is already actively queued (much like Add() on the Interface)
+//   - Delayed items are inserted into the 'waiting' queue according to their Waiting option (TakeShorter by default)
+//   - Immediate items will be dropped when the TakeLonger Waiting option is set and they are already present in 'waiting'.
+//   - Immediate items will pre-empt/remove items in the 'waiting' queue unless the PermitActiveAndWaiting option is set.
+//
+// processArrivingMessage will return entries
+func (q *delayingType) processArrivingMessage(w *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) {
+	// handle the different message actions
+	switch entry.action {
+	case waitForActionExists:
+		if entry.returnCh == nil {
+			return
+		}
+		r := 0
+		_, exists := knownEntries[entry.data]
+		if exists {
+			r = 1
+		}
+		if q.logger != nil {
+			q.log(fmt.Sprintf("waitingLoop: return exists result %t for %v", exists, entry.data))
+		}
+		entry.returnCh <- r
+		close(entry.returnCh)
+		return
+	case waitForActionLength:
+		if entry.returnCh == nil {
+			return
+		}
+		r := len(knownEntries)
+		if q.logger != nil {
+			q.log("waitingLoop: waiting queue length: " + strconv.Itoa(r))
+		}
+		entry.returnCh <- r
+		close(entry.returnCh)
```

Comment:
not a real benefit, but the following reason
- best practice to ensure that the close will be called as soon as we know we have the channel after `if entry.returnCh == nil`
- if there is a bug in the logger, the consumer will not be blocked and have their channel closed immediately 
- safer in case additional code will be added in the future

Reply:
ok so rather than the defer which will wait till the function return I should endeavour to put the logging after the close?  I can do that.  If the log was to hang then this would still stop the delaying queue in its tracks - can you think of a way I could prevent this?  Would it look something like a logging go-routine with a large feeder channel like the waitingLoop?

Reply:
Logging is now removed.